### PR TITLE
Fix delete of functions that becomes unused

### PR DESCRIPTION
**Pull request by hewj03**
After the first loop of deleting instructions in ValuesToDelete, deleted
instructions in ValuesToDelete are in an unstable state. Then in the
second loop of deleting, dyn_cast to GlobalValue could return true for
an instruction and double eraseFromParent causes crash. The crash is
observed in a multi-threaded test case.

Global values in ValuesToDelete are functions. Unused functions are
deleted by eraseUselessFunctions anyway.

### CODE REVIEW

**Comment by hewj03** [comment]
```
/home/runner/work/SPIRV-LLVM-Translator/SPIRV-LLVM-Translator/SPIRV-LLVM-Translator/lib/SPIRV/SPIRVReader.cpp:3207:54: error: ‘bool llvm::LLVMContext::supportsTypedPointers() const’ is deprecated: Always returns false [-Werror=deprecated-declarations]
 3207 |     UseTargetTypes = !Context->supportsTypedPointers();
```
CI failure isn't related to this PR.

**Review by MrSidims — APPROVED** [review]

File: lib/SPIRV/OCLToSPIRV.cpp [hunk 1]
```diff
@@ -179,9 +179,6 @@ bool OCLToSPIRVBase::runOCLToSPIRV(Module &Module) {
   for (auto &I : ValuesToDelete)
```
Comment (svenvh) [inline lib/SPIRV/OCLToSPIRV.cpp:179]: Perhaps we should change the type of `ValuesToDelete` to `std::set<Instruction *>` and avoid inserting the called functions into it?

Comment (hewj03) [reply lib/SPIRV/OCLToSPIRV.cpp]: done, thank you

**Comment by hewj03** [comment]
clang-format-17 is not found
```
Run cat diff-to-inspect.txt | /usr/share/clang/clang-format-18/clang-format-diff.py -p1 > clang-format.patch
Traceback (most recent call last):
  File "/usr/share/clang/clang-format-18/clang-format-diff.py", line [15](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/actions/runs/5745096659/job/15572569454?pr=2109#step:6:16)5, in main
    p = subprocess.Popen(
  File "/usr/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.8/subprocess.py", line [17](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/actions/runs/5745096659/job/15572569454?pr=2109#step:6:18)04, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'clang-format-17'
```

**Comment by hewj03** [comment]
The thread safety issue is observed on https://github.com/intel/intel-graphics-compiler/releases/tag/igc-1.0.14062.11 which uses llvm 14. Example of crash:
```
First loop
ValuesToDelete, eraseFromParent: 0x2871bd0
ValuesToDelete, eraseFromParent: 0x298e550
ValuesToDelete, eraseFromParent: 0x298f520
ValuesToDelete, eraseFromParent: 0x2994ee0
ValuesToDelete, eraseFromParent: 0x2995eb0
ValuesToDelete, eraseFromParent: 0x299b9e0
ValuesToDelete, eraseFromParent: 0x299c9b0
ValuesToDelete, eraseFromParent: 0x29bc940
ValuesToDelete, eraseFromParent: 0x29bed20
ValuesToDelete, eraseFromParent: 0x29fac80
ValuesToDelete, eraseFromParent: 0x2a459d0

Second loop:
ValuesToDelete, eraseFromParent:: 0x2839a48
ValuesToDelete, eraseFromParent:: 0x283a738
ValuesToDelete, eraseFromParent:: 0x283e8e8
ValuesToDelete, eraseFromParent:: 0x283f2d8
ValuesToDelete, eraseFromParent:: 0x2994ee0   [crash because 0x2994ee0 is already erased in the first loop]
```
I write a standalone test of llvm-spirv using similar code as in https://github.com/intel/opencl-clang/blob/e9d59b8525b657114951d89dd2da6d349f572bb7/opencl_clang.cpp#L294-L323. The test still crashes on llvm 14, but is working fine on llvm-15/llvm-16 branches. So it seems dyn_cast\<GlobalValue> is working fine in llvm 15/16 branches. Nevertheless, this PR should be still a refinement.

**Review by svenvh — APPROVED** [review]

**Comment by MrSidims** [comment]
>> clang-format-17 is not found

Guess that's excepted due to llvm upgrade, not all of packaged have yet updated

**Pull request merged**